### PR TITLE
docs(diagnostic): diagnostic-structure in the return value of get()

### LINKED ISSUE #24143: Documentation improvements for vim.diagnostic.get()
### Problem

TL;DL
Can you add that bufnr is not optional in the return value of `vim.diagnostic.get()`?

First, I apologize if my understanding is incorrect.

The return value of `vim.diagnostic.get()` is a list of `diagnostic-structure`.
```help
get({bufnr}, {opts})                                    *vim.diagnostic.get()*
    Get current diagnostics.

    Parameters: ~
      • {bufnr}  (integer|nil) Buffer number to get diagnostics from. Use 0
                 for current buffer or nil for all buffers.
      • {opts}   (table|nil) A table with the following keys:
                 • namespace: (number) Limit diagnostics to the given
                   namespace.
                 • lnum: (number) Limit diagnostics to the given line number.
                 • severity: See |diagnostic-severity|.

    Return: ~
        Diagnostic [] table A list of diagnostic items |diagnostic-structure|.
```

Help says `bufnr` of `diagnostic-structure` is optional.
```help
                                                        *diagnostic-structure*
A diagnostic is a Lua table with the following keys. Required keys are
indicated with (+):

    bufnr: Buffer number
    lnum(+): The starting line of the diagnostic
    end_lnum: The final line of the diagnostic
    col(+): The starting column of the diagnostic
    end_col: The final column of the diagnostic
    severity: The severity of the diagnostic |vim.diagnostic.severity|
    message(+): The diagnostic text
    source: The source of the diagnostic
    code: The diagnostic code
    user_data: Arbitrary data plugins or users can add

Diagnostics use the same indexing as the rest of the Nvim API (i.e. 0-based
rows and columns). |api-indexing|
```

### Expected behavior

`vim.diagnostic.set()`, however, fills bufnr and then creates a cache.
https://github.com/neovim/neovim/blob/cc624fac683ce6dbd01f10781817cd654cceed38/runtime/lua/vim/diagnostic.lua#L702
https://github.com/neovim/neovim/blob/cc624fac683ce6dbd01f10781817cd654cceed38/runtime/lua/vim/diagnostic.lua#L282-L282

It seems that caches cannot be added except by `set()`, and I think it is safe to assume that anything you can get with `get()` will always have bufnr in it.
So I would like to see this documented.

### PR DESCRIPTION
**Pull request by uga-rosa**
fix #24143

### CODE REVIEW

**Comment by uga-rosa** [comment]
I am not a native English speaker. Please feel free to point out any strange expressions.

**Comment by uga-rosa** [comment]
A test is failed, but I am only updating the document. It seems to be timing out on windows.

File: runtime/doc/diagnostic.txt [hunk 1]
```diff
@@ -53,6 +53,9 @@ indicated with (+):
     code: The diagnostic code
     user_data: Arbitrary data plugins or users can add
 
+In the return value of |vim.diagnostic.get()|, keys `bufnr`, `end_lnum`,
```
Comment (gpanders) [inline runtime/doc/diagnostic.txt:56]: The wording here is ok, but I would put this in the docstring of vim.diagnostic.get directly (not that this requires modifying the comment in diagnostic.lua, then regenerating the docs with the gen_vimdoc.py script).

Comment (uga-rosa) [reply runtime/doc/diagnostic.txt:56]: The text is long and there are not enough line breaks.

```diff
diff --git a/runtime/doc/diagnostic.txt b/runtime/doc/diagnostic.txt
index b12a506..b6596ec 100644
--- a/runtime/doc/diagnostic.txt
+++ b/runtime/doc/diagnostic.txt
@@ -53,9 +53,6 @@ indicated with (+):
     code: The diagnostic code
     user_data: Arbitrary data plugins or users can add

-In the return value of |vim.diagnostic.get()|, keys `bufnr`, `end_lnum`,
-`end_col`, and `severity` are guaranteed to be present.
-
 Diagnostics use the same indexing as the rest of the Nvim API (i.e. 0-based
 rows and columns). |api-indexing|

@@ -481,7 +478,8 @@ get({bufnr}, {opts})                                    *vim.diagnostic.get()*
                  • severity: See |diagnostic-severity|.

     Return: ~
-        Diagnostic [] table A list of diagnostic items |diagnostic-structure|.
+        Diagnostic [] table A list of diagnostic items |diagnostic-structure|. In the
+        return value of |vim.diagnostic.get()|, keys `bufnr` , `end_lnum` , `end_col` , and `severity` are guaranteed to be present.

 get_namespace({namespace})                    *vim.diagnostic.get_namespace()*
     Get namespace metadata.
diff --git a/runtime/lua/vim/diagnostic.lua b/runtime/lua/vim/diagnostic.lua
index 0d1d01b..4cfa0f6 100644
--- a/runtime/lua/vim/diagnostic.lua
+++ b/runtime/lua/vim/diagnostic.lua
@@ -767,7 +767,7 @@ end
 ---                        - namespace: (number) Limit diagnostics to the given namespace.
 ---                        - lnum: (number) Limit diagnostics to the given line number.
 ---                        - severity: See |diagnostic-severity|.
----@return Diagnostic[] table A list of diagnostic items |diagnostic-structure|.
+---@return Diagnostic[] table A list of diagnostic items |diagnostic-structure|. In the return value of |vim.diagnostic.get()|, keys `bufnr`, `end_lnum`, `end_col`, and `severity` are guaranteed to be present.
 function M.get(bufnr, opts)
   vim.validate({
     bufnr = { bufnr, 'n', true },
```

**Comment by justinmk** [comment]
> The text is long and there are not enough line breaks.

if the gen_vimdoc.py has a generation bug, just ignore it. It can be fixed later.

File: runtime/doc/diagnostic.txt (hunk 1 above)
Comment (gpanders) [reply runtime/doc/diagnostic.txt:56]: Remove this change and then it looks good to me.

Comment (justinmk) [reply runtime/doc/diagnostic.txt:56]: still waiting for this to be removed, before merging

**Comment by uga-rosa** [comment]
Sorry. For some reason I couldn't force push and it took a while.

File: runtime/doc/diagnostic.txt [hunk 2]
```diff
@@ -53,6 +53,8 @@ indicated with (+):
     code: The diagnostic code
     user_data: Arbitrary data plugins or users can add
 
+`end_col`, and `severity` are guaranteed to be present.
+
```
Comment (justinmk) [inline runtime/doc/diagnostic.txt:57]:
```suggestion
```

**Pull request merged**